File: ValidAnagram/ValidAnagram.java
package ValidAnagram;

import java.util.HashMap;
import java.util.Map;

public class ValidAnagram {
    public static void main(String[] args) {
        Solution s = new Solution();

        System.err.println("\n\n");
        System.out.println(s.isAnagram("anagram", "nagaram")); // esperado: true
        System.out.println(s.isAnagram("rat", "car")); // esperado: false
        System.out.println(s.isAnagram("aab", "abb")); // esperado: false
    }
}

class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) { return false; }
        int len = t.length();

        //map palavra s
        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < len; i++) { //percorre palavra
            Character c = s.charAt(i);

            if (map.containsKey(c)) { //se ja tem o char, qtd +1
                int value = map.get(c) + 1;
                map.put(c,value);
            } else {
                map.put(c, 1); //Se nao, qtd = 1
            }
        }

        //map palavra t
        for (int i = 0; i < len; i++) {
            Character c = t.charAt(i);
            if (map.containsKey(c)) {
                int value = map.get(c) - 1;
                map.put(c,value);
                
                if (value < 0) { return false; }
            } else {
                return false;
            }
        }
        
        return true;
    }
}

//Mapear palavra 1 (char, qtd)
//Mapear palavra 2 (char, qtd)
//percorrer os maps e ver se sao iguais. ?
//Percorro uma palavra e comparo os maps nos chars. Se der a mesma qtd ate o final -> true 

File: ValidAnagram/ValidAnagram.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

bool isAnagram(char* s, char* t) {
    int lenS = strlen(s);
    int lenT = strlen(t);

    if (lenS != lenT) { return false; }

    int count[26] = {0}; //para alfabeto

    for (int i = 0; i < lenS; i++) {
        count[s[i] - 'a']++; //incrementa no indice letra s[i] - 'a', conta feito em ASCII.
        count[t[i] - 'a']--; //decrementa na segunda palavra
    }

    for (int i = 0; i < 26; i++) { //se tiver algum indice != 0 -> Nao é anagrama
        if (count[i] != 0) { return false; }
    }

    return true;
}

int main() {
    printf("%d\n", isAnagram("anagram", "nagaram")); // esperado: 1 (true)
    printf("%d\n", isAnagram("rat", "car"));          // esperado: 0 (false)
    printf("%d\n", isAnagram("aab", "abb"));           // esperado: 0 (false)

    return 0;
}